src/Workspaces/Core/Portable/Workspace/Host/Metadata/MetadataReferenceCache.cs

```diff
@@ -0,0 +1,95 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Composition;
+using System.Diagnostics;
+using System.Linq;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Host
+{
+    /// <summary>
+    /// A cache for metadata references.
+    /// </summary>
+    internal class MetadataReferenceCache
+    {
+        private ImmutableDictionary<string, ReferenceSet> _referenceSets
+            = ImmutableDictionary<string, ReferenceSet>.Empty;
+
+        private readonly Func<string, MetadataReferenceProperties, MetadataReference> _createReference;
+
+        public MetadataReferenceCache(Func<string, MetadataReferenceProperties, MetadataReference> createReference)
+        {
+            if (createReference == null)
+            {
+                throw new ArgumentNullException(nameof(createReference));
+            }
+
+            this._createReference = createReference;
+        }
+
+        public MetadataReference GetReference(string path, MetadataReferenceProperties properties)
+        {
+            ReferenceSet referenceSet;
+            if (!_referenceSets.TryGetValue(path, out referenceSet))
+            {
+                referenceSet = ImmutableInterlocked.GetOrAdd(ref _referenceSets, path, new ReferenceSet(this));
+            }
+
+            return referenceSet.GetAddOrUpdate(path, properties);
+        }
+
+        /// <summary>
+        /// A collection of references to the same underlying metadata, each with different properties.
+        /// </summary>
+        private class ReferenceSet
+        {
+            private readonly MetadataReferenceCache _cache;
+
+            private readonly NonReentrantLock _gate = new NonReentrantLock();
```

Comment:
Minor tweak: pass "useThisInstanceForSynchronization: true" to the NonReentrantLock constructor to save an allocation here.